Refactoring in the skmultiflow.trees module

smastelini opened the pull request:
This PR addresses Issue #148. 

Due to the large number of necessary changes in the PR, I would like to request some help to verify and discuss the validity of the performed operations.

The main idea in this PR is to increase modularity and the organization in the skmultiflow.trees modules. To this end, I envisioned the tree implementations as five main elements:

- The tree core codes, per se
- Their nodes (split nodes, prediction nodes, and so forth)
- Their attribute observers (which monitor the attribute/target combinations to enable split attempts)
- The chosen split criterion to growth
- The types of attribute tests performed (binary tests, for instance)

As a result, I've created subfolders for the last 4 aspects, respectively, `nodes`, `attribute_observer`, `split_criterion`, and `attribute_test`. These names are just a sketch. Any suggestions for a better nomenclature are welcomed!

The needed changes for `attribute_observer`, `split_criterion`, and `attribute_test` were pretty straightforward. I had just to update some imports and configure/create some `__init__.py` files.
The changes for the leaf predictors, on the other hand, caused more impact. To decouple the inner classes from the trees, I had to create a lot of new files (one for each kind of leaf predictor, plus some basic node structures from `HoeffdingTree`). 

As I mentioned in Issue #148, I believe that there might be some redundant code among the plethora of leaf node classes. This might be a task for future improvements in the framework. Additionally, I believe that such refactoring of the tree module could help with Issue #138. Maybe we could extend the renaming process to the leaf nodes as well. I've tried to standardize the names of the new files as much as I could. I had to rename some classes to comply with the name of the trees they belong to. 

Nevertheless, there might be still some inconsistencies in the names. Unlike the tree algorithms (Issue #138), perhaps we could standardize the leaves names in this PR, since the users do not access them directly via the API. There would be no necessity to issue deprecation warnings in this case (IMHO :-D).  What do you think, @jacobmontiel? (On the other hand, the contributors would have to adapt to the new changes)

I also updated the documentation to reflect the changes. I noticed that some of the leaf nodes do not have (a complete) documentation. With the refactoring process, this became more apparent. We could either add the missing documentation in this PR or create some issues about this need. Any opinions on that matter?

I also edited some of the code to comply with PEP8. The changes were mostly some line breaks for very long lines.

As much as all the tests are passing (at least in my local machine), given the extent of the changes, I would appreciate some help in verifying possible missing/incorrect imports. It is also possible that there are unnecessary imports I may not have noticed.

It is worth noting that we could apply the same refactoring to the `skmultiflow.rules` module.

To finalize, I would like to thank you @jacobmontiel, @hmgomes and the skmultiflow community in advance for the help and patience during this PR.

Any comments and/or requests for changes are welcomed. I believe that the code at its current state is the first step to achieve the desired goals for this PR.

Cheers, and sorry for the long description.
Saulo

Changes proposed in this pull request:

* To streamline the Hoeffding Tree algorithms separating their leaf nodes, attribute observers, split criteria, and attribute tests.
* Improvements in some `__init__.py` files (adding missing exports for some classes/functions)
* Updates in the documentation
* Addresses #148.

---

Checklist

- [x] Code complies with PEP-8 and is consistent with the framework.
- [ ] Code is properly documented.
- [x] Tests are included for new functionality or updated accordingly.
- [x] Travis CI build passes with no errors.
- [ ] Test Coverage is maintained (threshold is -0.2%).
- [x] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

jacobmontiel commented:
Fail in Travis is due to changes in their environments. I updated the configuration to account for this and added a fix for `test_calculate_object_size` which fails in macOS.

Please merge the last commits in the master.

jacobmontiel commented on src/skmultiflow/trees/nodes/inactive_learning_node_perceptron_multi_target.py:
```diff
@@ -0,0 +1,84 @@
+from skmultiflow.utils.utils import get_dimensions
+
+
+class InactiveLearningNodePerceptronMultiTarget(InactiveLearningNodePerceptron):
```
Please add a short description.

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented on src/skmultiflow/trees/nodes/lc_active_learning_node.py:
```diff
@@ -0,0 +1,33 @@
+NAIVE_BAYES_ADAPTIVE = 'nba'
+
+
+class LCActiveLearningNode(ActiveLearningNode):
```
Please add a short description.

smastelini replied:
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

jacobmontiel commented on src/skmultiflow/trees/nodes/lc_inactive_learning_node.py:
```diff
@@ -0,0 +1,17 @@
+from skmultiflow.trees.nodes import InactiveLearningNode
+
+
+class LCInactiveLearningNode(InactiveLearningNode):
```
Please add a short description.

smastelini replied:
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

jacobmontiel commented on src/skmultiflow/trees/nodes/lc_learning_node_nb.py:
```diff
@@ -0,0 +1,46 @@
+from skmultiflow.trees.nodes import LCActiveLearningNode
+
+
+class LCLearningNodeNB(LCActiveLearningNode):
```
Please add a short description.

smastelini replied:
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

jacobmontiel commented on src/skmultiflow/trees/nodes/lc_learning_node_nba.py:
```diff
@@ -0,0 +1,65 @@
+from skmultiflow.trees.nodes import LCLearningNodeNB
+
+
+class LCLearningNodeNBA(LCLearningNodeNB):
```
Please add a short description.

smastelini replied:
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

jacobmontiel commented on src/skmultiflow/trees/nodes/sst_active_learning_node.py:
```diff
@@ -0,0 +1,157 @@
+from skmultiflow.utils.utils import get_dimensions
+
+
+class SSTActiveLearningNode(ActiveLearningNodePerceptronMultiTarget):
```
Please add a short description.

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented on src/skmultiflow/trees/nodes/sst_active_learning_node_adaptive.py:
```diff
@@ -0,0 +1,95 @@
+from skmultiflow.utils import get_dimensions
+
+
+class SSTActiveLearningNodeAdaptive(SSTActiveLearningNode):
```
Please add a short description.

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented on src/skmultiflow/trees/nodes/sst_inactive_learning_node.py:
```diff
@@ -0,0 +1,102 @@
+from skmultiflow.utils import get_dimensions
+
+
+class SSTInactiveLearningNode(InactiveLearningNodePerceptronMultiTarget):
```
Please add a short description.

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented on src/skmultiflow/trees/nodes/sst_inactive_learning_node_adaptive.py:
```diff
@@ -0,0 +1,95 @@
+from skmultiflow.utils.utils import get_dimensions
+
+
+class SSTInactiveLearningNodeAdaptive(SSTInactiveLearningNode):
```
Please add a short description.

smastelini replied:
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented on src/skmultiflow/trees/regression_hoeffding_adaptive_tree.py:
```diff
@@ -1,16 +1,14 @@
 
 _TARGET_MEAN = 'mean'
 _PERCEPTRON = 'perceptron'
 error_width_threshold = 300
```
Use uppercase for constants.

smastelini replied:
Change addressed in commit [a8c8022](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/a8c8022987ea6ac45b8aea5fa0cc2a1ab2b20384).

jacobmontiel commented on src/skmultiflow/trees/split_criterion/intra_cluster_variance_reduction_split_criterion.py:
```diff
@@ -37,4 +36,4 @@ def compute_SD(dist):
         sum_sq = dist[2]
 
-        return np.sum((sum_sq - (sum * sum) / N) / (N - 1))
+        return np.mean((sum_sq - ((sum * sum) / N)) / (N - 1))
```
Was this a bug? We need to document the rationale for this change and its impact.

smastelini replied:
Yes, it was. I had forgotten to create a PR for this aspect.

I found this bug while performing some experiments. As much as proportionally the values are similar, this indeed might impact in the ICVR calculation.

For clearness, I will submit a separate minor PR for this aspect. What do you think?

 How do I must proceed to document this aspect? Adding it to the PR description?
Thank you for pointing it out.

jacobmontiel replied:
This has been addressed in its own PR #152.

smastelini commented:
> Fail in Travis is due to changes in their environments. I updated the configuration to account for this and added a fix for `test_calculate_object_size` which fails in macOS.
> 
> Please merge the last commits in the master.

I just add the mentioned fix in the branch. Thank you!

smastelini commented on src/skmultiflow/trees/__init__.py:
```diff
@@ -2,14 +2,17 @@
 from .hoeffding_tree import HoeffdingTree
 from .hoeffding_adaptive_tree import HAT
 from .hoeffding_anytime_tree import HATT
+from .arf_hoeffding_tree import ARFHoeffdingTree
```
I agree with you. When I was updating the `__init__.py` file I figured out that `HATT` was not being "exposed" so I changed it and also put `ARFHoeffdingTree`. This latter one has not direct purpose in being "indexed" in this way. I will commit a change concerning this matter.

smastelini replied:
This was addressed in a commit.

smastelini commented on src/skmultiflow/trees/__init__.py:
```diff
@@ -2,14 +2,17 @@
 The :mod:`skmultiflow.trees` module includes learning methods based on trees.
 """
 
+from .attribute_split_suggestion import AttributeSplitSuggestion
```
Agreed. I will address this in a commit.

smastelini replied:
It was addressed in a commit.

smastelini commented on src/skmultiflow/trees/arf_hoeffding_tree.py:
```diff
@@ -286,21 +109,28 @@ def __init__(self,
         if self._leaf_prediction == MAJORITY_CLASS:
-            return self.RandomLearningNode(initial_class_observations, self.max_features,
-                                           random_state=self._random_state)
+            return ARFLearningNode(
```
The main purpose was trying to find a way to tie the node implementation with the trees that use them.

As I previously mentioned:

> I've tried to standardize the names of the new files as much as I could. I had to rename some classes to comply with the name of the trees they belong to.

> Nevertheless, there might be still some inconsistencies in the names. Unlike the tree algorithms (Issue #138), perhaps we could standardize the leaves names in this PR, since the users do not access them directly via the API.

Maybe we could use something like `RandomLearningNodeClassification` or some other variation that might make clearer the node's purpose. Comments?

jacobmontiel replied:
This is indeed a tricky topic. Using the names of the trees could be confusing if we want to re-use them in different methods (e.g. `ActiveLearningNode` from the Hoeffding Tree). I like the idea of stating the type of learning (classification or regression). `RandomLearningNodeClassification` seems like a good compromise.

Comments @hmgomes?

smastelini replied:
I have renamed the `ARFLearningNode*` classes to `RandomLearningNode*`.

smastelini commented on src/skmultiflow/trees/nodes/ada_node_regression.py:
```diff
@@ -0,0 +1,37 @@
+from abc import ABCMeta, abstractmethod
+
+
+class AdaNodeRegression(metaclass=ABCMeta):
```
This is a good point.

In this initial effort a basically just decoupled the leaves implementations from their trees. As I pointed out previously:

> As I mentioned in Issue #148, I believe that there might be some redundant code among the plethora of leaf node classes.

This seems to be the case here. Thank you for pointing it out. I am going to verify this fact and use the same abstract class if possible.

smastelini replied:
I have removed the `AdaNodeRegression` class in commit [014137b](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/014137b147b8858d95aac977ead44a5a569f55d0). All of its references were changed to `AdaNode`.

smastelini commented on src/skmultiflow/trees/nodes/anytime_active_learning_node.py:
```diff
@@ -0,0 +1,84 @@
+
+
+class AnyTimeActiveLearningNode(ActiveLearningNode):
+    """ Learning node that supports growth.
```
I agree. I am going to create a commit addressing the missing/incomplete/wrong documentation for the leaf nodes.

smastelini replied:
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/hoeffding_anytime_tree.py:
```diff
@@ -71,449 +67,22 @@ class HATT(HoeffdingTree):
+            return AnyTimeActiveLearningNode(initial_class_observations)
         elif self._leaf_prediction == NAIVE_BAYES:
-            return self.AnyTimeLearningNodeNB(initial_class_observations)
+            return AnyTimeLearningNodeNB(initial_class_observations)
```
Where should I add it?

In the documentation `nba` already appears as the default option for `leaf_prediction`.
Suggestions?

jacobmontiel replied:
We can edit the existing comment before the `else:`

`# NAIVE BAYES ADAPTIVE (default)`

This is to minimize the chances of future errors from people unfamiliar with the code.

smastelini replied:
Addressed by commit [95988c5](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/95988c58446d70b7a166b14f2281aa40c86075ab).

jacobmontiel replied:
Hi @smastelini. I think that this was not included in the referenced commit.

smastelini commented on src/skmultiflow/trees/lc_hoeffding_tree.py:
```diff
@@ -293,26 +155,26 @@ def _new_learning_node(self, initial_class_observations=None):
+            return LCLearningNodeNB(initial_class_observations)
         else:
-            return self.LCLearningNodeNBA(initial_class_observations)
+            return LCLearningNodeNBA(initial_class_observations)
```
I think it lies in the same situation of `HoeffdingAnytimeTree`.

jacobmontiel replied:
Agree, the same suggestion applies. In this case, we need to add the comments for each clause.

smastelini replied:
Addressed by commit [95988c5](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/95988c58446d70b7a166b14f2281aa40c86075ab).

smastelini commented on src/skmultiflow/trees/multi_target_regression_hoeffding_tree.py:
```diff
@@ -71,356 +78,6 @@ class MultiTargetRegressionHoeffdingTree(RegressionHoeffdingTree, MultiOutputMix
-            )
-
     # ===========================================
     # == Hoeffding Regression Tree implementation ===
```
Comment correct in a commit.

smastelini commented on src/skmultiflow/trees/nodes/__init__.py:
```diff
@@ -0,0 +1,61 @@
+"""
+The :mod:`skmultiflow.trees.leaf_predictor` module includes learning methods
```
Package description updated in a recent commit.

smastelini commented on src/skmultiflow/trees/regression_hoeffding_adaptive_tree.py:
```diff
@@ -60,258 +58,6 @@ class RegressionHAT(RegressionHoeffdingTree):
-            found_nodes.append(HoeffdingTree.FoundNode(self, parent, parent_branch))
-
     # ===========================================
     # == Hoeffding Regression Tree implementation ===
```
Comment corrected in a commit.

smastelini commented on src/skmultiflow/trees/nodes/ada_split_node_for_regression.py:
```diff
@@ -0,0 +1,163 @@
+from skmultiflow.drift_detection.adwin import ADWIN
+from skmultiflow.utils import check_random_state
+
+error_width_threshold = 300
```
Change addressed in commit [a8c8022](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/a8c8022987ea6ac45b8aea5fa0cc2a1ab2b20384).

smastelini commented on src/skmultiflow/trees/nodes/ada_learning_node.py:
```diff
@@ -0,0 +1,110 @@
+MAJORITY_CLASS = 'mc'
+NAIVE_BAYES = 'nb'
+NAIVE_BAYES_ADAPTIVE = 'nba'
+error_width_threshold = 300
```
Change addressed in commit [a8c8022](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/a8c8022987ea6ac45b8aea5fa0cc2a1ab2b20384).

smastelini commented on src/skmultiflow/trees/nodes/ada_split_node.py:
```diff
@@ -0,0 +1,157 @@
+from skmultiflow.drift_detection import ADWIN
+from skmultiflow.utils import check_random_state, get_max_value_key
+
+error_width_threshold = 300
```
Change addressed in commit [a8c8022](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/a8c8022987ea6ac45b8aea5fa0cc2a1ab2b20384).

smastelini commented:
I have already addressed some of the suggested changes:

- Removing the export for `AttributeSplitSuggestion` and `ARFHoeffdingTree` from `skmultiflow.trees.__init__.py`
- Renaming the 'ARFLearningNode*' classes to `RandomLearningNode*`
- Updating `skmultiflow.trees.nodes` description
- Using uppercase for constants in the Adaptive Hoeffding Tree implementations
- Correcting outdated comments and documentation in the impacted code
- Removing unnecessary `AdaNodeRegression` abstract class (using `AdaNode` instead)

My next step will be to create missing documentation for the learning nodes.
Nonetheless, there is seem to be some problem with the building process. Any thoughts in that matter @jacobmontiel?

smastelini commented on src/skmultiflow/trees/nodes/active_learning_node_adaptive_multi_target.py:
```diff
@@ -0,0 +1,58 @@
+from skmultiflow.trees.nodes import ActiveLearningNodePerceptronMultiTarget
+
+
+class ActiveLearningNodeAdaptiveMultiTarget(ActiveLearningNodePerceptronMultiTarget):
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/active_learning_node_for_regression.py:
```diff
@@ -0,0 +1,59 @@
+
+
+class ActiveLearningNodeForRegression(ActiveLearningNode):
+
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/active_learning_node_for_regression_multi_target.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class ActiveLearningNodeForRegressionMultiTarget(ActiveLearningNodeForRegression):
+
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/active_learning_node_perceptron.py:
```diff
@@ -0,0 +1,112 @@
+
+
+class ActiveLearningNodePerceptron(ActiveLearningNode):
+
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/active_learning_node_perceptron_multi_target.py:
```diff
@@ -0,0 +1,131 @@
+
+
+class ActiveLearningNodePerceptronMultiTarget(ActiveLearningNodePerceptron):
+
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/inactive_learning_node_adaptive_multi_target.py:
```diff
@@ -0,0 +1,51 @@
+from skmultiflow.trees.nodes import InactiveLearningNodePerceptronMultiTarget
+
+
+class InactiveLearningNodeAdaptiveMultiTarget(InactiveLearningNodePerceptronMultiTarget):
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/inactive_learning_node_for_regression.py:
```diff
@@ -0,0 +1,19 @@
+from skmultiflow.trees.nodes import InactiveLearningNode
+
+
+class InactiveLearningNodeForRegression(InactiveLearningNode):
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

smastelini commented on src/skmultiflow/trees/nodes/inactive_learning_node_perceptron.py:
```diff
@@ -0,0 +1,49 @@
+from skmultiflow.trees.nodes import InactiveLearningNode
+
+
+class InactiveLearningNodePerceptron(InactiveLearningNode):
```
Documentation updated in commit [5b724c9](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5b724c98a7bcc2d270f1034185b2978201235576).

jacobmontiel commented:
Seems like the culprit is the last commit. 

All errors (`ImportError: cannot import name 'attribute_observer'`) are triggered by this line:

https://github.com/scikit-multiflow/scikit-multiflow/blob/f3fb6c26ef2c38c9f8f422aa376ffa7a3ea2f0df/src/skmultiflow/trees/__init__.py#L5

I suspect the other lines on that commit will result in a similar import error.

jacobmontiel commented:
I have finished a second past on the code. Thanks for the changes. I have marked as resolved the corresponding items.

smastelini commented:
> > Nonetheless, there is seem to be some problem with the building process.
> 
> Seems like the culprit is the last commit.
> 
> All errors (`ImportError: cannot import name 'attribute_observer'`) are triggered by this line:
> 
> https://github.com/scikit-multiflow/scikit-multiflow/blob/f3fb6c26ef2c38c9f8f422aa376ffa7a3ea2f0df/src/skmultiflow/trees/__init__.py#L5
> 
> I suspect the other lines on that commit will result in a similar import error.

Thank you for the feedback.

I was able to fix the building problems. I had to add a new `setup.py` file in the `skmultiflow.trees` module. I followed the same pattern used in the `lazy` submodule and in `skmultiflow.setup.py`.

What do you think @jacobmontiel? Do you suggest changes in the way I handled it?

jacobmontiel commented:
> I was able to fix the building problems. I had to add a new `setup.py` file in the `skmultiflow.trees` module. I followed the same pattern used in the `lazy` submodule and in `skmultiflow.setup.py`.

Great! Now I understand what happened. My bad, I failed to communicate that change in the setup process. Previously, subpackages were detected automatically. As the framework grows we need to have better control over its structure. This is why now all subpackages (folders) are now manually added in the configuration.

smastelini commented on src/skmultiflow/trees/nodes/ada_learning_node.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class AdaLearningNode(LearningNodeNBAdaptive, AdaNode):
+    def __init__(self, initial_class_observations):
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/ada_learning_node_for_regression.py:
```diff
@@ -0,0 +1,72 @@
+
+
+class AdaLearningNodeForRegression(ActiveLearningNodePerceptron, AdaNodeRegression):
+
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/ada_split_node.py:
```diff
@@ -0,0 +1,157 @@
+error_width_threshold = 300
+
+
+class AdaSplitNode(SplitNode, AdaNode):
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/ada_split_node_for_regression.py:
```diff
@@ -0,0 +1,163 @@
+error_width_threshold = 300
+
+
+class AdaSplitNodeForRegression(SplitNode, AdaNodeRegression):
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/anytime_inactive_learning_node.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class AnyTimeInactiveLearningNode(InactiveLearningNode):
+    """ Inactive learning node that does not grow.
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/anytime_learning_node_nb.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class AnyTimeLearningNodeNB(AnyTimeActiveLearningNode):
+
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented on src/skmultiflow/trees/nodes/anytime_learning_node_nb_adaptive.py:
```diff
@@ -0,0 +1,61 @@
+from skmultiflow.trees.nodes import AnyTimeLearningNodeNB
+
+
+class AnyTimeLearningNodeNBAdaptive(AnyTimeLearningNodeNB):
```
Description added in commit [5f331dc](https://github.com/scikit-multiflow/scikit-multiflow/pull/149/commits/5f331dce48293665f7ef78b059ba2d6782606438).

smastelini commented:
I just addressed the initial revisions proposed for this PR. This last step referred to adding missing descriptions for some of the leaf node implementations.

Unfortunately, they are in some sense generic, since I do not know most of Hoeffding Tree variants in depth. As much as I added descriptions for the nodes, some of their methods also lack proper documentation. 
This is something where some help would be largely welcomed, given my lack of knowledge on the specifics of the algorithms. What do you think @jacobmontiel, should we open an issue asking for help to document the mentioned methods?

My next two main concerns about this PR are the following:

- It would be nice to adopt a standardized nomenclature for the leaf nodes. Ex: we have the active/inactive prefixes, but some implementations just use `*LearningNode*` (without indicating whether they are active or inactive). Most of the node implementations also do not indicate in their names whether they are suited to classification or regression tasks. Another important point to discuss refer to the nodes named after the HT variants.

- I have the feeling that some portions of the leaf node implementations are redundant. Perhaps with proper inheritance usage, we could alleviate this problem. Maybe this could be something to deal with in the future. Any thoughts @jacobmontiel?

Besides these two points, I'm open to any further suggestions for this PR. Thank you, @jacobmontiel, for your support so far!

jacobmontiel commented:
I think those are valid concerns but fall outside of the scope of the PR. The changes and improved documentation are a step forward. I vote to create issues to consider in future development.

> It would be nice to adopt a standardized nomenclature for the leaf nodes.

Agree. One issue could focus on standardization and documentation.

> I have the feeling that some portions of the leaf node implementations are redundant.

This can be a different issue focused on the implementation details. This is something worth considering and we could also take the chance to explore ways to optimize performance.

smastelini commented:
> Seems that some nodes are not covered by current tests, thus now that they are in new files the diff is dropping. I suggest including some coverage tests by setting different leaves. However, I am open to creating a new issue to address this.

I am currently implementing a "pilot" test for LCHT to see how the coverage is impacted.

Is there an easy way of detecting which specific node implementation miss proper tests?

jacobmontiel commented:
> Is there an easy way of detecting which specific node implementation miss proper tests?

Yes, you can use the report from `codecov` available [here](https://codecov.io/gh/scikit-multiflow/scikit-multiflow/compare/68c68464311ca90667db9644f8b700317ea5ecc1...5f331dce48293665f7ef78b059ba2d6782606438/diff)

Although the simplest solution is to run the test with coverage locally:

```bash
$ pytest tests/trees -v --cov=src/skmultiflow/trees --cov-report html
```

This will run **only** the tests for `trees` and an html report will be created. You can use that report to *see* coverage.

**Note:** You will need to install [pytest-cov](https://github.com/pytest-dev/pytest-cov) to run the above-mentioned command.

smastelini commented:
Thank you again, @jacobmontiel. I ran the tests locally using the coverage reports. From them, I could add some additional tests in the last three commits. They cover different leaf predictors for the impacted HT variants and also the Gini split criterion. The coverage increased a little bit, but the `diff hit` is still below the target.

Using the coverage report, I could identify the main currently weaknesses for the tests (IMHO):

- Inactive learning mostly are not covered (this brings up again the discussions in Issue #128) 
    * I believe that implementing tests to cover nodes in an inactive state (for memory-constrained situations), as well as addressing #128 could be a milestone for the future.

- Hoeffding Adaptive Tree variants also lack proper coverage. In this case, it would be needed to simulate concept drifts to make the tests hitting uncovered portions of the code.

How do you think those points fit in the context of this PR? Are they objectives to pursue here or address in the future?

jacobmontiel commented:
> How do you think those points fit in the context of this PR? Are they objectives to pursue here or address in the future?

I think these points fall outside the scope of the PR. We could use the same issue mentioned before for revisiting the implementation to also explore better coverage. Since the nodes are now outside of the trees we could test them independently. As you mention, the trickiest part of testing trees is to meet the required conditions. I am just trying to figure out how to make our life easier. Note, this does not mean that we shouldn't perform sanity checks in the tree variants that use them. Could you create new issues from our discussion?

I am ok with the PR as it is, the new structure is ready and documentation/testing is improved. If you agree, I will close the PR and merge the changes once the new issues are created and referenced.

Thanks again for the hard work!

smastelini commented:
> I am ok with the PR as it is, the new structure is ready and documentation/testing is improved. If you agree, I will close the PR and merge the changes once the new issues are created and referenced.

To reflect the points raised in this PR, I've just created Issues #154, #155, #156, and #157. Their descriptions might be improved. For instance, we could analyze and depict which exact learning nodes need proper documentation (Issue #156). I can edit the description of the Issues in the future to make them clear, if necessary.

Did I miss something?

jacobmontiel commented:
The PR is ready to close. Changes are correct. New issues have been created for items outside the scope of this PR. Thanks @smastelini!!!
